options/options.go

```diff
@@ -0,0 +1,218 @@
+// Copyright OpenSSF Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package options
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/caarlos0/env/v6"
+
+	"github.com/ossf/scorecard-action/github"
+	"github.com/ossf/scorecard/v4/options"
+)
+
+var (
+	// Errors.
+	errGithubEventPathEmpty       = errors.New("GitHub event path is empty")
+	errResultsPathEmpty           = errors.New("results path is empty")
+	errOnlyDefaultBranchSupported = errors.New("only default branch is supported")
+
+	trueStr = "true"
+)
+
+// Options are options for running scorecard via GitHub Actions.
+type Options struct {
+	// Scorecard options.
+	ScorecardOpts *options.Options
+
+	// Scorecard command-line options.
+	EnabledChecks string `env:"ENABLED_CHECKS"`
+
+	// Scorecard checks.
+	EnableLicense           string `env:"ENABLE_LICENSE"`
+	EnableDangerousWorkflow string `env:"ENABLE_DANGEROUS_WORKFLOW"`
+
+	// GitHub options.
+	// TODO(github): Consider making this a separate options set so we can
+	//               encapsulate handling
+	GithubEventName  string `env:"GITHUB_EVENT_NAME"`
+	GithubEventPath  string `env:"GITHUB_EVENT_PATH"`
+	GithubRef        string `env:"GITHUB_REF"`
+	GithubRepository string `env:"GITHUB_REPOSITORY"`
+	GithubWorkspace  string `env:"GITHUB_WORKSPACE"`
+
+	DefaultBranch string `env:"SCORECARD_DEFAULT_BRANCH"`
+	// TODO(options): This may be better as a bool
+	IsForkStr string `env:"SCORECARD_IS_FORK"`
+	// TODO(options): This may be better as a bool
+	PrivateRepoStr string `env:"SCORECARD_PRIVATE_REPOSITORY"`
+}
+
+const (
+	defaultScorecardPolicyFile = "policy.yml"
+	formatSarif                = options.FormatSarif
+)
+
+// New creates a new options set for running scorecard via GitHub Actions.
+func New() (*Options, error) {
+	// Enable scorecard command to use SARIF format.
+	os.Setenv(options.EnvVarEnableSarif, trueStr)
+
+	opts := &Options{
+		ScorecardOpts: options.New(),
+	}
+	if err := env.Parse(opts); err != nil {
+		return opts, fmt.Errorf("parsing entrypoint env vars: %w", err)
+	}
+
+	if err := opts.Initialize(); err != nil {
+		return opts, fmt.Errorf(
+			"initializing scorecard-action options: %w",
+			err,
+		)
+	}
+
+	// TODO(options): Move this set-or-default logic to its own function.
+	opts.ScorecardOpts.EnableSarif = true
+	if opts.ScorecardOpts.Format == formatSarif {
```

Comment:
Nit: use a single `if` instead of nested for better readability.

Reply:
> Nit: use a single `if` instead of nested for better readability.

@azeemshaikh38 -- I wrestled with this a little...

- `scorecard` currently defaults to `FormatDefault` (which I believe is the table writer)
- using `scorecard.options.New()` will get you the default values of the options, but can be overridden during initialization with env vars, like `SCORECARD_RESULTS_FORMAT`
- `scorecard-action` really wants a default format of `options.FormatSarif`, which means:
- you also need to ensure that `ENABLE_SARIF` is set in the enviroment (because the SARIF option is feature-flagged)
- only when SARIF format is chosen AND `ENABLE_SARIF` is set will `PolicyFile` be available as a CLI flag, which means:
- you should only expose the option to set `PolicyFile` when you know SARIF is enabled

I asked in the channel, but I'll ask here too... do we want not bother exposing the format here altogether and just default to SARIF for `scorecard-action`?

Reply:
> do we want not bother exposing the format here altogether and just default to SARIF for `scorecard-action`?

I think having SARIF as the default format in `scorecard-action` is fine, but users may want to dump out a JSON instead of SARIF?

Even though I can't think of a usecase where someone is using JSON instead of SARIF, the work involved in un-exposing the format option (we'll need to make sure current users aren't broken) is far greater than the benefit it gets us.

Reply:
> I think having SARIF as the default format in `scorecard-action` is fine, but users may want to dump out a JSON instead of SARIF?
> 
> Even though I can't think of a usecase where someone is using JSON instead of SARIF, the work involved in un-exposing the format option (we'll need to make sure current users aren't broken) is far greater than the benefit it gets us.

Yep, this would be dramatically simpler without the env vars, etc., but I want to ensure backwards-compat, so I agree that it's worth keeping both.